indico/modules/events/editing/client/js/components/FileManager/FileList.jsx

```diff
@@ -0,0 +1,152 @@
+// This file is part of Indico.
+// Copyright (C) 2002 - 2019 CERN
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import React, {useContext, useRef, useState} from 'react';
+import PropTypes from 'prop-types';
+import Dropzone from 'react-dropzone';
+import {Icon} from 'semantic-ui-react';
+import {FileManagerContext, filePropTypes, uploadFiles, deleteFile} from './util';
+import * as actions from './actions';
+
+import './FileManager.module.scss';
+
+function FileAction({onClick, active, icon, className}) {
+  return (
+    <Icon
+      className={className}
+      name={active ? 'spinner' : icon}
+      loading={active}
+      onClick={() => {
+        if (!active) {
+          onClick();
+        }
+      }}
+    />
+  );
+}
+
+FileAction.propTypes = {
+  onClick: PropTypes.func.isRequired,
+  active: PropTypes.bool.isRequired,
+  icon: PropTypes.string.isRequired,
+  className: PropTypes.string.isRequired,
+};
+
+function FileEntry({
+  eventId,
+  fileTypeId,
+  allowMultipleFiles,
+  file: {uuid, filename, state, claimed},
+}) {
+  const ref = useRef(null);
+  const dispatch = useContext(FileManagerContext);
+  const [activeButton, setActiveButton] = useState(null);
+
+  return (
+    <>
+      <span styleName="file-state" className={state || ''}>
+        {filename}
+      </span>
+      <span>
+        {!state && allowMultipleFiles && (
+          <>
+            <FileAction
+              icon="exchange"
+              active={activeButton === 'replace'}
+              styleName="exchange-icon"
+              onClick={() => {
+                ref.current.open();
+              }}
+            />
+            <Dropzone
+              ref={ref}
+              onDrop={async acceptedFiles => {
+                setActiveButton('replace');
+                await uploadFiles(
+                  actions.markModified,
+                  fileTypeId,
+                  acceptedFiles,
+                  eventId,
+                  dispatch,
+                  uuid
+                );
+              }}
+            >
+              {({getRootProps, getInputProps}) => (
+                <span {...getRootProps()}>
+                  <input {...getInputProps()} />
+                </span>
+              )}
+            </Dropzone>
+          </>
+        )}
+        {state !== 'deleted' && state !== 'modified' ? (
+          <FileAction
+            styleName="delete-icon"
+            icon="trash"
+            active={activeButton === 'delete'}
+            onClick={async () => {
+              if (!claimed) {
+                // if the file is not part of the revision yet,
+                // we can safely delete it from the server
+                setActiveButton('delete');
+                await deleteFile(uuid);
```

Comment:
Deleting the file is more a "nice to do" thing, so we could immediately delete it without waiting for the request. Anyway, nothing wrong with doing it this way.

Reply:
Yeah, I thought of that, but anyway, doesn't hurt to wait...